runtime/src/snapshot_utils.rs

```diff
@@ -1769,6 +1803,102 @@ fn rebuild_bank_from_snapshots(
     Ok(bank)
 }
 
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+fn verify_slot_deltas(
+    slot_deltas: &[BankSlotDelta],
+    bank: &Bank,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    verify_slot_deltas_structural(slot_deltas, bank.slot())?;
+    verify_slot_deltas_with_history(slot_deltas, &bank.get_slot_history(), bank.slot())
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks are simple/structural
+fn verify_slot_deltas_structural(
+    slot_deltas: &[BankSlotDelta],
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // there should not be more entries than that status cache's max
+    let num_entries = slot_deltas.len();
+    if num_entries > status_cache::MAX_CACHE_ENTRIES {
+        return Err(VerifySlotDeltasError::TooManyEntries(
+            num_entries,
+            status_cache::MAX_CACHE_ENTRIES,
+        ));
+    }
+
+    let mut slots_seen_so_far = HashSet::new();
+    for &(slot, is_root, ..) in slot_deltas {
+        // all entries should be roots
+        if !is_root {
+            return Err(VerifySlotDeltasError::SlotIsNotRoot(slot));
+        }
+
+        // all entries should be for slots less than or equal to the bank's slot
+        if slot > bank_slot {
+            return Err(VerifySlotDeltasError::SlotGreaterThanMaxRoot(
+                slot, bank_slot,
+            ));
+        }
+
+        // there should only be one entry per slot
+        let inserted = slots_seen_so_far.insert(slot);
+        if !inserted {
+            return Err(VerifySlotDeltasError::SlotHasMultipleEntries(slot));
+        }
+    }
+
+    Ok(())
+}
+
+/// Verify that the snapshot's slot deltas are not corrupt/invalid
+/// These checks use the slot history for verification
+fn verify_slot_deltas_with_history(
+    slot_deltas: &[BankSlotDelta],
+    slot_history: &SlotHistory,
+    bank_slot: Slot,
+) -> std::result::Result<(), VerifySlotDeltasError> {
+    // all entries should be in the bank's slot history
+    let entry_missing_from_history = slot_deltas
+        .iter()
+        .find(|(slot, ..)| slot_history.check(*slot) != Check::Found);
+    if let Some((slot, ..)) = entry_missing_from_history {
+        return Err(VerifySlotDeltasError::SlotNotFoundInHistory(*slot));
+    }
+
+    // all slots in the history should be in the slot deltas (up to MAX_CACHE_ENTRIES)
+    // this ensures nothing was removed from the status cache
+
+    fn slot_deltas_contains_slot(slot_deltas: &[BankSlotDelta], slot: Slot) -> bool {
+        slot_deltas
+            .iter()
+            .any(|(slot_from_deltas, ..)| *slot_from_deltas == slot)
+    }
+
+    // go through the slot history and make sure there's an entry for each slot
+    // note: it's important to go highest-to-lowest since the status cache is a fifo
+    let mut entries_checked = 0;
+    for slot in (0..=bank_slot).rev() {
```

Comment:
how about using `slot_history.oldest()`?:

there's the very unlikely hypothetical bug when bank_slot is large and its slot_history is crafted so that only the bank_slot is Found. In that case, this loop doesn't terminate way down until to reach to the slot 0. The reason is that `entries_checked` isn't incremented in that situation.

instead, using `slot_history.oldest()` limits this iteration up to `SlotHistory::MAX_ENTRIES`, which isn't controllable by the attacker, unlike the sysvar.

note that you can't trust sysvars at this phase of snapshot verification, unfortunately. you can only verify that the slot history sysvar and status cache are at least consistent and retains integrity.

oh, it looks like my iterator version does suffer from this bug too.

that said, the final form would be like this:

```rust
    if slot_history.oldest() > bank_slot {
      return Err(BadSlotHistory)
    }
    for slot in ... {
        ...
        return Err(VerifySlotDeltasError::SlotNotFoundInDeltas(slot));
        ...
    }
    // detect premature loop termination.
    if slot != slot_history.oldest() && entries_checked < MAX_CACHE_ENTRIES {
        return Err(TooFewEntries) // actually, we're now verification the exact lower bound of the count of slot deltas, being counterpart of TooManyEntries.
    }
```

Reply:
> Though I'm not sure I see a need for the "detect premature loop termination" check...

never mind, I was forgetting the existence of `SlotNotFoundInHistory`... now, i think your conclusion is right.